### DISCUSSION TITLE
Upgrading from 11.3.1 to 12.1.0

### DISCUSSION THREAD

**Discussion by 31337** [docker-mailserver/docker-mailserver#3415] [General]
I don't feel safe to upgrade from 11.3.1 to 12.1.0 because I can't seem to find a kind of "howto" do it.
There are several changelogs, documentation updates here and there, changes to different configuration files but no "when upgrading from 11.3.1 to 12, check the following list: ...".

I want to enable RSPAMD in 12.1.0.

What I did so far:

- rename docker-compose.yml to compose.yaml
- in compose.yaml update URL to ghcr.io/docker-mailserver/docker-mailserver:12.1.0

- in mailserver.env: ENABLE_AMAVIS=0, ENABLE_SPAMASSASIN=0

- downloaded new setup.sh

If I pull 12.1.0 now will everything just work out of the box?
What about DKIM/DMARC? Is this going to continue to work just like with 11.3.1?

If I missed some upgrade-documentation, please point me to it! Thanks.

Edit: I'm using OPENDKIM at the moment. Can I reuse the existing keys for RSPAMD by just copying them over to the RSPAM directory, or do they have to be re-generated? Or is it recommended to continue to use OPENDKIM?
For OPENDMARC I would just disable it. (ENABLE_OPENDMARC=0)

**Comment by polarathene** [comment]
> * downloaded new `setup.sh`

This is not necessary these days. You can use `docker exec` / `docker compose exec` to get into the container and run the `setup` command which supports the same features. The `setup.sh` script is mostly for some minor convenience now and runs the same `setup` commands within the container.

---

> If I pull 12.1.0 now will everything just work out of the box?

Hard to say.

Some users had non-standard configuration where they replaced the Dovecot config file we updated in v12. This caused problems for them until they applied the same changes for SASL.

---

> What about DKIM/DMARC? Is this going to continue to work just like with 11.3.1?

It has had some changes, but from what I've heard RSPAMD differs a little as well. You could first try with the default OpenDKIM / OpenDMARC that you'd already have, and later switch?

I was not involved much in the rspamd work, and the main maintainer for this is quite busy for a while. I recall more work going on to improve the rspamd support with v13, you may want to stick with what has been working for you so far? It'll be switched to default in a future major version.

---

> If I missed some upgrade-documentation, please point me to it! Thanks.

The release notes sometimes differ from the [Changelog file](https://github.com/docker-mailserver/docker-mailserver/blob/master/CHANGELOG.md), but most changes are documented there for awareness. And provide reference links to work on those changes which can go into much more technical detail.

We don't tend to have migration/upgrade guides beyond that. We're rather limited on time and do our best to document what we can and raise awareness of anything critical. Other than that you can try looking at bug reports (some closed and resolved). Not everything is added to release notes or changelog, but we do tag the version to related issues / work, such as:

- [`milestone:v12.1.0`](https://github.com/docker-mailserver/docker-mailserver/pulls?q=milestone%3Av12.1.0+)
- [`milestone:v12.0.0`](https://github.com/docker-mailserver/docker-mailserver/pulls?q=milestone%3Av12.0.0+)
- Likewise for [an upcoming `13`](https://github.com/docker-mailserver/docker-mailserver/issues?q=milestone%3Av13.0.0+), might reveal some bugs / fixes that are queued up.

---

> I'm using OPENDKIM at the moment. Can I reuse the existing keys for RSPAMD by just copying them over to the RSPAM directory, or do they have to be re-generated? Or is it recommended to continue to use OPENDKIM?

I think the keys are compatible. We have a way to generate keys from each tool which will output slightly differently, but that's not an issue IIRC. However rspamd requires extra configuration (as described in the docs), so you might want to just keep it simple by regenerating, or if viable, send a test mail with the existing keys in use, and if that works you should be all good.

I don't think there's any issue with using opendkim. One of the other maintainers still uses that without issues AFAIK (_I don't use DMS myself_).

**Comment by 31337** [comment]
Thanks for the answer. I upgraded to 12.1.0 now, using ENABLE_OPENDKIM=0. In rspamd you can use the old opendkim keys, I just had to rename them and put them in the according rspamd directory. DKIM seems to work fine with rspamd. Rspamd itself is a different beast. Had to (wanted to) disable a few modules enabled by default (like RBL). Still fighting a bit with the configuration, have to read the Rspamd docs a bit I guess. In general the upgrade went through fine.

**Comment by carlos1680** [comment]
Hello, I upgraded to last version but stop working.
Just, run pull command, stop y start compose file but doesnt send or receive emails now.
In log file, there is a message similar to:
[private/dovecot-lmtp]: No such file or directory)
The only specific thing I use is dmark, after that is standard for me.
Could you help me?
Thanks.
Carlos.-

**Reply by casperklein** [reply]
Destroy the container completely (`docker-compose down; docker-compose up --force-recreate`) and try again. If that doesn't solve your problem, please open a new issue and provide information about your setup.

**Reply by polarathene** [reply]
> `[private/dovecot-lmtp]: No such file or directory`

This is a v12 migration issue. The Dovecot LMTP socket was moved, however that is only expected to break if you've got a custom Dovecot config file mounted previously as that contains old config instead of the updated config from the v12 release. Several users ran into this issue. 

Otherwise, force recreating the container like advised earlier should do the trick. If not, please do open a proper bug report, share your config and reference this discussion.

**Reply by carlos1680** [reply]
Hola, gracias por la respuesta. intentare las recomendaciones.
Saludos.
Carlos
